File: research/experiments/mathexpr-ref-python-sonnet/mathexpr.py
"""
Math expression parser and evaluator.

A Type-O reference implementation translated from TypeScript.
Supports basic arithmetic with proper operator precedence.
"""

from dataclasses import dataclass
from typing import Literal, Union


# ============================================================================
# Token Types
# ============================================================================

TokenKind = Literal[
    "number", "plus", "minus", "star", "slash", "percent", "power", "lparen", "rparen"
]


@dataclass
class Token:
    """A lexical token from the input string."""
    kind: TokenKind
    value: str


def token(kind: TokenKind, value: str) -> Token:
    """Create a token with the given kind and value."""
    return Token(kind=kind, value=value)


# ============================================================================
# AST Types
# ============================================================================

BinaryOp = Literal["+", "-", "*", "/", "%", "**"]
UnaryOp = Literal["-"]


@dataclass
class NumberLiteral:
    """A numeric literal node."""
    type: Literal["number"] = "number"
    value: float = 0.0


@dataclass
class UnaryExpr:
    """A unary expression node."""
    type: Literal["unary"] = "unary"
    op: UnaryOp = "-"
    operand: "AstNode" = None


@dataclass
class BinaryExpr:
    """A binary expression node."""
    type: Literal["binary"] = "binary"
    op: BinaryOp = "+"
    left: "AstNode" = None
    right: "AstNode" = None


AstNode = Union[NumberLiteral, UnaryExpr, BinaryExpr]


def number_literal(value: float) -> NumberLiteral:
    """Create a number literal AST node."""
    return NumberLiteral(type="number", value=value)


def unary_expr(op: UnaryOp, operand: AstNode) -> UnaryExpr:
    """Create a unary expression AST node."""
    return UnaryExpr(type="unary", op=op, operand=operand)


def binary_expr(op: BinaryOp, left: AstNode, right: AstNode) -> BinaryExpr:
    """Create a binary expression AST node."""
    return BinaryExpr(type="binary", op=op, left=left, right=right)


# ============================================================================
# Tokenizer
# ============================================================================

def tokenize(input_str: str) -> list[Token]:
    """
    Convert a math expression string into a sequence of tokens.
    
    Supports: integers, decimals, operators (+, -, *, /, %, **), parentheses.
    Whitespace is skipped. Raises ValueError on unrecognized characters.
    """
    tokens: list[Token] = []
    i = 0

    while i < len(input_str):
        ch = input_str[i]

        # Skip whitespace
        if ch in " \t\n\r":
            i += 1
            continue

        # Parentheses
        if ch == "(":
            tokens.append(token("lparen", "("))
            i += 1
            continue

        if ch == ")":
            tokens.append(token("rparen", ")"))
            i += 1
            continue

        # Single-character operators
        if ch == "+":
            tokens.append(token("plus", "+"))
            i += 1
            continue

        if ch == "-":
            tokens.append(token("minus", "-"))
            i += 1
            continue

        if ch == "/":
            tokens.append(token("slash", "/"))
            i += 1
            continue

        if ch == "%":
            tokens.append(token("percent", "%"))
            i += 1
            continue

        # Star or power
        if ch == "*":
            if i + 1 < len(input_str) and input_str[i + 1] == "*":
                tokens.append(token("power", "**"))
                i += 2
            else:
                tokens.append(token("star", "*"))
                i += 1
            continue

        # Numbers (including decimals)
        if ch.isdigit() or ch == ".":
            num = ""
            has_dot = False
            while i < len(input_str) and (input_str[i].isdigit() or input_str[i] == "."):
                if input_str[i] == ".":
                    if has_dot:
                        raise ValueError(f"Unexpected character '.' at position {i}")
                    has_dot = True
                num += input_str[i]
                i += 1
            tokens.append(token("number", num))
            continue

        # Unrecognized character
        raise ValueError(f"Unexpected character '{ch}' at position {i}")

    return tokens


# ============================================================================
# Parser
# ============================================================================

def parse(tokens: list[Token]) -> AstNode:
    """
    Parse a sequence of tokens into an AST using recursive descent.
    
    Operator precedence (lowest to highest):
      1. Addition, subtraction (+, -)
      2. Multiplication, division, modulo (*, /, %)
      3. Exponentiation (**)  — right-associative
      4. Unary minus (-)
      5. Atoms: numbers, parenthesized expressions
    """
    pos = 0

    def peek() -> Token | None:
        """Get current token without advancing."""
        return tokens[pos] if pos < len(tokens) else None

    def advance() -> Token:
        """Consume and return current token."""
        nonlocal pos
        t = tokens[pos]
        pos += 1
        return t

    def expect(kind: TokenKind) -> Token:
        """Consume a token of the expected kind, or raise an error."""
        t = peek()
        if t is None or t.kind != kind:
            raise ValueError(f"Expected {kind} but got {t.kind if t else 'end of input'}")
        return advance()

    # Level 1: addition and subtraction (lowest precedence)
    def parse_add_sub() -> AstNode:
        left = parse_mul_div()
        while peek() and peek().kind in ("plus", "minus"):
            op_token = advance()
            op = "+" if op_token.kind == "plus" else "-"
            right = parse_mul_div()
            left = binary_expr(op, left, right)
        return left

    # Level 2: multiplication, division, modulo
    def parse_mul_div() -> AstNode:
        left = parse_power()
        while peek() and peek().kind in ("star", "slash", "percent"):
            op_token = advance()
            if op_token.kind == "star":
                op = "*"
            elif op_token.kind == "slash":
                op = "/"
            else:
                op = "%"
            right = parse_power()
            left = binary_expr(op, left, right)
        return left

    # Level 3: exponentiation (right-associative)
    def parse_power() -> AstNode:
        base = parse_unary()
        if peek() and peek().kind == "power":
            advance()
            exponent = parse_power()  # right-recursive for right-associativity
            return binary_expr("**", base, exponent)
        return base

    # Level 4: unary minus
    def parse_unary() -> AstNode:
        if peek() and peek().kind == "minus":
            advance()
            operand = parse_unary()  # allow chained unary: --x
            return unary_expr("-", operand)
        return parse_atom()

    # Level 5: atoms — numbers and parenthesized expressions
    def parse_atom() -> AstNode:
        t = peek()

        if t is None:
            raise ValueError("Unexpected end of input")

        if t.kind == "number":
            advance()
            return number_literal(float(t.value))

        if t.kind == "lparen":
            advance()
            expr = parse_add_sub()
            expect("rparen")
            return expr

        raise ValueError(f"Unexpected token: {t.kind} '{t.value}'")

    # Parse the expression
    ast = parse_add_sub()

    # Ensure all tokens were consumed
    if pos < len(tokens):
        remaining = tokens[pos]
        raise ValueError(f"Unexpected token after expression: {remaining.kind} '{remaining.value}'")

    return ast


# ============================================================================
# Evaluator
# ============================================================================

def evaluate(node: AstNode) -> float:
    """
    Evaluate an AST node to produce a numeric result.
    
    Supports: +, -, *, /, %, ** (power), unary negation.
    Raises ValueError on division by zero and modulo by zero.
    """
    if node.type == "number":
        return node.value

    if node.type == "unary":
        return -evaluate(node.operand)

    # Binary expression
    left = evaluate(node.left)
    right = evaluate(node.right)

    if node.op == "+":
        return left + right
    if node.op == "-":
        return left - right
    if node.op == "*":
        return left * right
    if node.op == "**":
        return left ** right

    if node.op == "/":
        if right == 0:
            raise ValueError("Division by zero")
        return left / right

    # node.op == "%"
    if right == 0:
        raise ValueError("Modulo by zero")
    return left % right


# ============================================================================
# Public API
# ============================================================================

def calc(expression: str) -> float:
    """
    End-to-end expression evaluation: string in, number out.
    
    Composes tokenizer → parser → evaluator into a single function.
    This is the root node of the library — the public API.
    """
    if expression.strip() == "":
        raise ValueError("Empty expression")
    
    tokens = tokenize(expression)
    ast = parse(tokens)
    return evaluate(ast)

File: research/experiments/mathexpr-ref-go-sonnet/mathexpr.go
package mathexpr

import (
	"fmt"
	"math"
	"strconv"
)

// TokenKind represents the type of a token
type TokenKind string

const (
	TokenNumber  TokenKind = "number"
	TokenPlus    TokenKind = "plus"
	TokenMinus   TokenKind = "minus"
	TokenStar    TokenKind = "star"
	TokenSlash   TokenKind = "slash"
	TokenPercent TokenKind = "percent"
	TokenPower   TokenKind = "power"
	TokenLParen  TokenKind = "lparen"
	TokenRParen  TokenKind = "rparen"
)

// Token represents a lexical token
type Token struct {
	Kind  TokenKind
	Value string
}

// NewToken creates a new token
func NewToken(kind TokenKind, value string) Token {
	return Token{Kind: kind, Value: value}
}

// AstNode is the interface for all AST node types
type AstNode interface {
	astNode()
}

// NumberLiteral represents a numeric literal
type NumberLiteral struct {
	Value float64
}

func (NumberLiteral) astNode() {}

// UnaryExpr represents a unary expression
type UnaryExpr struct {
	Op      string
	Operand AstNode
}

func (UnaryExpr) astNode() {}

// BinaryExpr represents a binary expression
type BinaryExpr struct {
	Op    string
	Left  AstNode
	Right AstNode
}

func (BinaryExpr) astNode() {}

// NewNumberLiteral creates a number literal node
func NewNumberLiteral(value float64) NumberLiteral {
	return NumberLiteral{Value: value}
}

// NewUnaryExpr creates a unary expression node
func NewUnaryExpr(op string, operand AstNode) UnaryExpr {
	return UnaryExpr{Op: op, Operand: operand}
}

// NewBinaryExpr creates a binary expression node
func NewBinaryExpr(op string, left, right AstNode) BinaryExpr {
	return BinaryExpr{Op: op, Left: left, Right: right}
}

// Tokenize converts a math expression string into a sequence of tokens
func Tokenize(input string) ([]Token, error) {
	tokens := []Token{}
	i := 0

	for i < len(input) {
		ch := input[i]

		// Skip whitespace
		if ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r' {
			i++
			continue
		}

		// Left paren
		if ch == '(' {
			tokens = append(tokens, NewToken(TokenLParen, "("))
			i++
			continue
		}

		// Right paren
		if ch == ')' {
			tokens = append(tokens, NewToken(TokenRParen, ")"))
			i++
			continue
		}

		// Plus
		if ch == '+' {
			tokens = append(tokens, NewToken(TokenPlus, "+"))
			i++
			continue
		}

		// Minus
		if ch == '-' {
			tokens = append(tokens, NewToken(TokenMinus, "-"))
			i++
			continue
		}

		// Star or Power
		if ch == '*' {
			if i+1 < len(input) && input[i+1] == '*' {
				tokens = append(tokens, NewToken(TokenPower, "**"))
				i += 2
			} else {
				tokens = append(tokens, NewToken(TokenStar, "*"))
				i++
			}
			continue
		}

		// Slash
		if ch == '/' {
			tokens = append(tokens, NewToken(TokenSlash, "/"))
			i++
			continue
		}

		// Percent
		if ch == '%' {
			tokens = append(tokens, NewToken(TokenPercent, "%"))
			i++
			continue
		}

		// Number
		if isDigit(ch) || ch == '.' {
			num := ""
			hasDot := false
			for i < len(input) && (isDigit(input[i]) || input[i] == '.') {
				if input[i] == '.' {
					if hasDot {
						return nil, fmt.Errorf("Unexpected character '.' at position %d", i)
					}
					hasDot = true
				}
				num += string(input[i])
				i++
			}
			tokens = append(tokens, NewToken(TokenNumber, num))
			continue
		}

		return nil, fmt.Errorf("Unexpected character '%c' at position %d", ch, i)
	}

	return tokens, nil
}

func isDigit(ch byte) bool {
	return ch >= '0' && ch <= '9'
}

// Parse parses a sequence of tokens into an AST
func Parse(tokens []Token) (AstNode, error) {
	pos := 0

	var peek func() *Token
	var advance func() Token
	var expect func(TokenKind) (Token, error)
	var parseAddSub func() (AstNode, error)
	var parseMulDiv func() (AstNode, error)
	var parsePower func() (AstNode, error)
	var parseUnary func() (AstNode, error)
	var parseAtom func() (AstNode, error)

	peek = func() *Token {
		if pos < len(tokens) {
			return &tokens[pos]
		}
		return nil
	}

	advance = func() Token {
		t := tokens[pos]
		pos++
		return t
	}

	expect = func(kind TokenKind) (Token, error) {
		t := peek()
		if t == nil {
			return Token{}, fmt.Errorf("Expected %s but got end of input", kind)
		}
		if t.Kind != kind {
			return Token{}, fmt.Errorf("Expected %s but got %s", kind, t.Kind)
		}
		return advance(), nil
	}

	// Level 1: addition and subtraction (lowest precedence)
	parseAddSub = func() (AstNode, error) {
		left, err := parseMulDiv()
		if err != nil {
			return nil, err
		}
		for {
			t := peek()
			if t == nil || (t.Kind != TokenPlus && t.Kind != TokenMinus) {
				break
			}
			opToken := advance()
			op := "+"
			if opToken.Kind == TokenMinus {
				op = "-"
			}
			right, err := parseMulDiv()
			if err != nil {
				return nil, err
			}
			left = NewBinaryExpr(op, left, right)
		}
		return left, nil
	}

	// Level 2: multiplication, division, modulo
	parseMulDiv = func() (AstNode, error) {
		left, err := parsePower()
		if err != nil {
			return nil, err
		}
		for {
			t := peek()
			if t == nil || (t.Kind != TokenStar && t.Kind != TokenSlash && t.Kind != TokenPercent) {
				break
			}
			opToken := advance()
			op := "*"
			if opToken.Kind == TokenSlash {
				op = "/"
			} else if opToken.Kind == TokenPercent {
				op = "%"
			}
			right, err := parsePower()
			if err != nil {
				return nil, err
			}
			left = NewBinaryExpr(op, left, right)
		}
		return left, nil
	}

	// Level 3: exponentiation (right-associative)
	parsePower = func() (AstNode, error) {
		base, err := parseUnary()
		if err != nil {
			return nil, err
		}
		t := peek()
		if t != nil && t.Kind == TokenPower {
			advance()
			exponent, err := parsePower() // right-recursive for right-associativity
			if err != nil {
				return nil, err
			}
			return NewBinaryExpr("**", base, exponent), nil
		}
		return base, nil
	}

	// Level 4: unary minus
	parseUnary = func() (AstNode, error) {
		t := peek()
		if t != nil && t.Kind == TokenMinus {
			advance()
			operand, err := parseUnary() // allow chained unary: --x
			if err != nil {
				return nil, err
			}
			return NewUnaryExpr("-", operand), nil
		}
		return parseAtom()
	}

	// Level 5: atoms — numbers and parenthesized expressions
	parseAtom = func() (AstNode, error) {
		t := peek()
		if t == nil {
			return nil, fmt.Errorf("Unexpected end of input")
		}

		if t.Kind == TokenNumber {
			advance()
			value, err := strconv.ParseFloat(t.Value, 64)
			if err != nil {
				return nil, err
			}
			return NewNumberLiteral(value), nil
		}

		if t.Kind == TokenLParen {
			advance()
			expr, err := parseAddSub()
			if err != nil {
				return nil, err
			}
			_, err = expect(TokenRParen)
			if err != nil {
				return nil, err
			}
			return expr, nil
		}

		return nil, fmt.Errorf("Unexpected token: %s '%s'", t.Kind, t.Value)
	}

	ast, err := parseAddSub()
	if err != nil {
		return nil, err
	}

	if pos < len(tokens) {
		remaining := tokens[pos]
		return nil, fmt.Errorf("Unexpected token after expression: %s '%s'", remaining.Kind, remaining.Value)
	}

	return ast, nil
}

// Evaluate evaluates an AST node to produce a numeric result
func Evaluate(node AstNode) (float64, error) {
	switch n := node.(type) {
	case NumberLiteral:
		return n.Value, nil

	case UnaryExpr:
		operand, err := Evaluate(n.Operand)
		if err != nil {
			return 0, err
		}
		return -operand, nil

	case BinaryExpr:
		left, err := Evaluate(n.Left)
		if err != nil {
			return 0, err
		}
		right, err := Evaluate(n.Right)
		if err != nil {
			return 0, err
		}

		switch n.Op {
		case "+":
			return left + right, nil
		case "-":
			return left - right, nil
		case "*":
			return left * right, nil
		case "**":
			return math.Pow(left, right), nil
		case "/":
			if right == 0 {
				return 0, fmt.Errorf("Division by zero")
			}
			return left / right, nil
		case "%":
			if right == 0 {
				return 0, fmt.Errorf("Modulo by zero")
			}
			return math.Mod(left, right), nil
		default:
			return 0, fmt.Errorf("Unknown operator: %s", n.Op)
		}

	default:
		return 0, fmt.Errorf("Unknown node type")
	}
}

// Calc is the public API function that evaluates a math expression string
func Calc(expression string) (float64, error) {
	if len(expression) == 0 || len(expression) == len(trimSpace(expression)) && trimSpace(expression) == "" {
		return 0, fmt.Errorf("Empty expression")
	}

	tokens, err := Tokenize(expression)
	if err != nil {
		return 0, err
	}

	ast, err := Parse(tokens)
	if err != nil {
		return 0, err
	}

	return Evaluate(ast)
}

func trimSpace(s string) string {
	start := 0
	end := len(s)
	for start < end && (s[start] == ' ' || s[start] == '\t' || s[start] == '\n' || s[start] == '\r') {
		start++
	}
	for end > start && (s[end-1] == ' ' || s[end-1] == '\t' || s[end-1] == '\n' || s[end-1] == '\r') {
		end--
	}
	return s[start:end]
}
